lib/core/src/Cardano/Wallet/Primitive/Types.hs

```diff
@@ -511,6 +517,32 @@ instance FromText PoolId where
         where
             textDecodingError = Left . TextDecodingError . show
 
+-- | Coins are stored as Lovelace (reminder: 1 Lovelace = 1e-6 ADA)
+newtype Stake = Stake { getStake :: Word64 } deriving (Show, Eq, Generic)
+
+data StakeDistribution = StakeDistribution
+    { dangling :: Stake
+    , pools :: [(PoolId,Stake)]
+    , unassigned :: Stake
+    } deriving (Eq, Show, Generic)
+
+instance FromJSON StakeDistribution where
+    parseJSON = genericParseJSON defaultOptions
+
+instance FromJSON Stake where
+    parseJSON (Number n) = case toBoundedInteger n of
+        Nothing ->
+            fail "stake should be non-negative integer"
+        Just stake ->
+            return (Stake stake)
+    parseJSON _ = fail "stake should be numeric"
+
+instance FromJSON PoolId where
+    parseJSON (String txt) = case fromText txt of
+        Left (TextDecodingError err) -> fail err
+        Right bs -> return bs
+    parseJSON _ = fail "stake pool id should be text"
```

Comment:
suggestion:

```hs
instance FromJSON PoolId where
    parseJSON = parseJSON >=> eitherToParser 
        . bimap ShowFmt (PoolId . T.encodeUtf8)
        . fromText
```

Reply:
this version will not check the proper stake pool id length, and the test : 
```
62          it "invalid stake pool id in endpoint response gives expected error" $ do                                                                                                         
63              let exampleStake = "{\"epoch\": 252054,\"stake\": {\"dangling\":0,\"\                                                                                                         
64                      \pools\":[[\"b80fed0d2289d535a94f6870add0cf8b3\                                                                                                                       
65                      \74cfe6cae078320ec\",1]],\"unassigned\":100100000000000}}"                                                                                                            
66              decodeJSON exampleStake `shouldBe`                                                                                                                                            
67                  Left "Error in $.stake.pools[0][0]: stake pool id invalid: \                                                                                                              
68                       \expected 32 bytes but got 25"                                                                                                                                       
69              return () 
```
will pass

Reply:
```hs
instance FromJSON PoolId where
    parseJSON = parseJSON >=> eitherToParser . left ShowFmt . fromText
```

:thinking: ?

Reply:
indeed, the following code works : 
```
207 instance FromJSON (ApiT PoolId) where                                                                                                                                                     
208     parseJSON val = do                                                                                                                                                                    
209         txt <- parseJSON val                                                                                                                                                              
210         m <- eitherToParser $ left ShowFmt $ fromText txt                                                                                                                                 
211         return $ ApiT m
```